File: 2024/02/day_02.jl
mutable struct Solution
    function Solution()
        new()
    end
end

function _evaluate(solution::Solution, numbers::Vector{Int})::Int
    rev = false

    for i in 2:length(numbers)
        if numbers[i] < numbers[i-1]
            rev = true
            break
        end
    end

    sorted = sort(numbers, rev=rev)
    spacing_okay = true

    for i in 2:length(numbers)
        if !(1 <= abs(numbers[i] - numbers[i-1]) <= 3)
            spacing_okay = false
            break
        end
    end

    if (sorted == numbers) && spacing_okay
        return 1
    end

    return 0
end

function evaluate(solution::Solution, line::String)::Tuple{Int, Int}
    numbers = map(x->parse(Int, x), split(line, " "))

    basic = _evaluate(solution, numbers)

    if basic == 1
        return (1, 1)
    end

    for i in eachindex(numbers)
        new_numbers = [numbers[j] for j in eachindex(numbers) if j != i]
        if _evaluate(solution, new_numbers) > 0
            return(0, 1)
        end
    end

    return (0, 0)
end

function read_data(fn::String)::Vector{String}
    return readlines(fn)
end

if abspath(PROGRAM_FILE) == @__FILE__
    data = read_data("input.txt")
    s = Solution()

    results = map(x->evaluate(s, x), data)

    safe = sum(first, results)
    dampened = sum(last, results)

    println(safe)
    println(dampened)
end

File: 2024/02/day_02.py
import argparse
from typing import Tuple, List

def _evaluate(rep: List[int]) -> int:
    asc = rep[1] > rep[0]

    if rep[0] == rep[1]:
        return 0
    
    last_elem = rep[0]

    for elem in rep[1:]:
        if asc and (elem <= last_elem or elem - last_elem > 3):
            return 0
        elif not asc and (elem >= last_elem or last_elem - elem> 3):
            return 0
        last_elem = elem
    return 1

def evaluate_new(report: str) -> Tuple[int, int]:
    rep = [int(x) for x in report.strip().split()]
    rv1 =  _evaluate(rep)
    if rv1 == 1:
        return (1,1)

    # brute force this
    rv2 = min(sum([_evaluate(rep[:x]+rep[x+1:]) for x in range(0, len(rep))]), 1)
    return rv1, rv2


def main(fn: str):
    with open(fn) as f:
        data = f.readlines()
    
    safe = 0
    dampened = 0
    for line in data:
        s, d = evaluate_new(line)
        safe = safe + s
        dampened = dampened + d
    print(f"there are {safe} safe reports")
    print(f"there are {dampened} safe dampened reports")


if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument("input_file", type=str)
    args = parser.parse_args()

    main(fn=args.input_file)